File: mobile/src/services/SocketService.ts
import {io, Socket} from 'socket.io-client';
import {Config} from '@/constants/config';
import AsyncStorage from '@react-native-async-storage/async-storage';

class SocketService {
  private socket: Socket | null = null;
  private listeners: Map<string, Set<Function>> = new Map();
  private lastError: string | null = null;
  private authSuccess = false;

  private log(...args: any[]) {
    console.log('[SocketService]', ...args);
  }

  getLastError(): string | null {
    return this.lastError;
  }

  isAuthenticated(): boolean {
    return this.authSuccess;
  }

  async connect() {
    if (this.socket?.connected) {
      this.log('Already connected, socket id:', this.socket.id);
      return;
    }

    const token = await AsyncStorage.getItem('access_token');
    this.log('Connecting with token:', token ? 'present' : 'missing');
    this.lastError = null;
    this.authSuccess = false;

    const isSecure = Config.WS_URL.startsWith('wss://') || Config.WS_URL.startsWith('https://');

    this.socket = io(Config.WS_URL, {
      path: '/socket.io',
      transports: ['websocket', 'polling'],
      auth: token ? {token} : undefined,
      query: token ? {token} : undefined,
      secure: isSecure,
      forceNew: true,
      timeout: 10000,
      reconnection: true,
      reconnectionAttempts: Infinity,
      reconnectionDelay: 1000,
      reconnectionDelayMax: 5000,
    });

    // Re-register all persistent listeners on the new socket instance.
    for (const [event, callbacks] of this.listeners) {
      for (const cb of callbacks) {
        const wrapper = (...args: any[]) => {
          try {
            this.log('RECEIVE', event, JSON.stringify(args));
          } catch {
            this.log('RECEIVE', event, '[non-serializable payload]');
          }
          try {
            cb(...args);
          } catch (err) {
            console.error(`[SocketService] listener error for ${event}:`, err);
          }
        };
        (cb as any).__wrapper = wrapper;
        this.socket.on(event, wrapper);
      }
    }

    this.socket.on('connect', () => {
      this.log('CONNECTED, socket id:', this.socket?.id);
      // Дублируем аутентификацию событием authenticate — как делает frontend.
      // Если токен не дошёл через auth/query, сервер сможет аутентифицировать здесь.
      if (token) {
        this.log('EMIT authenticate after connect');
        this.socket?.emit('authenticate', {access_token: token});
      }
    });

    this.socket.on('disconnect', (reason: string) => {
      this.log('DISCONNECTED, reason:', reason);
      this.authSuccess = false;
    });

    this.socket.on('connect_error', (err: any) => {
      const msg = err?.message || String(err);
      this.lastError = msg;
      this.log('CONNECT_ERROR:', msg, err);
    });

    this.socket.on('reconnect', (attempt: number) => {
      this.log('RECONNECTED after attempt:', attempt);
    });

    this.socket.on('reconnect_attempt', (attempt: number) => {
      this.log('RECONNECT_ATTEMPT:', attempt);
    });

    this.socket.on('reconnect_error', (err: any) => {
      const msg = err?.message || String(err);
      this.lastError = msg;
      this.log('RECONNECT_ERROR:', msg);
    });

    this.socket.on('error', (err: any) => {
      const msg = err?.message || String(err);
      this.lastError = msg;
      this.log('SOCKET_ERROR:', msg, err);
    });

    this.socket.on('connection_success', (data: any) => {
      this.log('CONNECTION_SUCCESS:', JSON.stringify(data));
      this.authSuccess = true;
      this.lastError = null;
    });

    // Presence / status events
    this.socket.on('user_status_change', (data: any) => {
      this.log('EVENT user_status_change:', JSON.stringify(data));
    });

    this.socket.on('presence_update', (data: any) => {
      this.log('EVENT presence_update:', JSON.stringify(data));
    });
  }

  getSocket(): Socket | null {
    return this.socket;
  }

  emit(event: string, data?: any) {
    this.log('EMIT', event, JSON.stringify(data));
    if (!this.socket) {
      this.log('EMIT FAILED: socket is null');
      return;
    }
    if (!this.socket.connected) {
      this.log('EMIT FAILED: socket not connected');
      return;
    }
    this.socket.emit(event, data);
  }

  on(event: string, callback: Function) {
    this.log('ON register:', event);
    if (!this.listeners.has(event)) {
      this.listeners.set(event, new Set());
    }
    this.listeners.get(event)!.add(callback);

    if (this.socket) {
      const wrapper = (...args: any[]) => {
        try {
          this.log('RECEIVE', event, JSON.stringify(args));
        } catch {
          this.log('RECEIVE', event, '[non-serializable payload]');
        }
        try {
          callback(...args);
        } catch (err) {
          console.error(`[SocketService] listener error for ${event}:`, err);
        }
      };
      (callback as any).__wrapper = wrapper;
      this.socket.on(event, wrapper);
    }
  }

  off(event: string, callback: Function) {
    this.log('OFF unregister:', event);
    if (this.listeners.has(event)) {
      this.listeners.get(event)!.delete(callback);
      if (this.listeners.get(event)!.size === 0) {
        this.listeners.delete(event);
      }
    }
    if (this.socket) {
      const wrapper = (callback as any).__wrapper;
      this.socket.off(event, wrapper || callback);
    }
  }

  disconnect() {
    this.log('DISCONNECT requested');
    this.authSuccess = false;
    this.socket?.disconnect();
    this.socket = null;
  }

  async reconnect() {
    this.log('RECONNECT requested');
    this.disconnect();
    await this.connect();
    // Re-register queued listeners
    for (const [event, callbacks] of this.listeners) {
      for (const cb of callbacks) {
        this.on(event, cb);
      }
    }
    this.listeners.clear();
  }
}

export const socketService = new SocketService();

File: mobile_flutter/lib/core/socket/socket_service.dart
import 'dart:convert';
import 'package:logger/logger.dart';
import 'package:socket_io_client/socket_io_client.dart' as IO;
import '../config/config.dart';
import '../utils/storage_service.dart';

class SocketService {
  final StorageService _storageService;
  final Logger _logger = Logger(printer: SimplePrinter(colors: true));
  
  IO.Socket? _socket;
  final Map<String, Set<Function>> _listeners = {};
  String? _lastError;
  bool _authSuccess = false;
  String? _currentToken;

  SocketService(this._storageService);

  String? getLastError() => _lastError;
  bool isAuthenticated() => _authSuccess;
  IO.Socket? getSocket() => _socket;

  Future<void> connect() async {
    final token = await _storageService.readSecure('access_token');

    if (_socket != null && _socket!.connected && _currentToken == token && _authSuccess) {
      _logger.d('[SocketService] Already connected and authenticated with current token');
      return;
    }

    _logger.d('[SocketService] Connecting. Token changed: ${_currentToken != token}, AuthSuccess: $_authSuccess');

    if (_socket != null) {
      disconnect();
    }

    _currentToken = token;
    _lastError = null;
    _authSuccess = false;

    final options = IO.OptionBuilder()
        .setPath('/socket.io')
        .setTransports(['websocket'])
        .enableForceNew()
        .enableAutoConnect()
        .setReconnectionDelay(1000)
        .setReconnectionDelayMax(5000)
        .setTimeout(10000);

    if (token != null) {
      options.setAuth({'token': token});
      options.setQuery({'token': token});
    }

    _socket = IO.io(AppConfig.wsUrl, options.build());

    // Re-register persistent listeners on the new socket
    _listeners.forEach((event, callbacks) {
      _socket!.on(event, (data) {
        _logger.d('[SocketService] RECEIVE $event: ${jsonEncode(data)}');
        final list = List<Function>.from(_listeners[event] ?? []);
        for (final cb in list) {
          try {
            cb(data);
          } catch (e) {
            _logger.e('[SocketService] Listener error for $event: $e');
          }
        }
      });
    });

    _socket!.onConnect((_) {
      _logger.d('[SocketService] CONNECTED, socket ID: ${_socket?.id}');
      if (token != null) {
        _logger.d('[SocketService] EMIT authenticate after connect');
        _socket?.emit('authenticate', {'access_token': token});
      }
    });

    _socket!.onDisconnect((reason) {
      _logger.d('[SocketService] DISCONNECTED, reason: $reason');
      _authSuccess = false;
    });

    _socket!.onConnectError((err) {
      _lastError = err?.toString();
      _logger.e('[SocketService] CONNECT_ERROR: $err');
    });

    _socket!.onError((err) {
      _lastError = err?.toString();
      _logger.e('[SocketService] SOCKET_ERROR: $err');
    });

    _socket!.on('connection_success', (data) {
      _logger.d('[SocketService] CONNECTION_SUCCESS: ${jsonEncode(data)}');
      _authSuccess = true;
      _lastError = null;
    });

    _socket!.on('user_status_changed', (data) {
      _logger.d('[SocketService] user_status_changed: ${jsonEncode(data)}');
    });

    _socket!.on('presence_update', (data) {
      _logger.d('[SocketService] presence_update: ${jsonEncode(data)}');
    });
  }

  void emit(String event, [dynamic data]) {
    if (_socket == null) {
      _logger.w('[SocketService] EMIT FAILED: socket is null');
      return;
    }
    if (!_socket!.connected) {
      _logger.w('[SocketService] EMIT FAILED: socket not connected');
      return;
    }
    _logger.d('[SocketService] EMIT $event: ${jsonEncode(data)}');
    _socket!.emit(event, data);
  }

  void on(String event, Function callback) {
    _logger.d('[SocketService] Register listener: $event');
    final callbacks = _listeners.putIfAbsent(event, () => <Function>{});
    final isNew = callbacks.isEmpty;
    callbacks.add(callback);

    if (_socket != null && isNew) {
      _socket!.on(event, (data) {
        _logger.d('[SocketService] RECEIVE $event: ${jsonEncode(data)}');
        final list = List<Function>.from(_listeners[event] ?? []);
        for (final cb in list) {
          try {
            cb(data);
          } catch (e) {
            _logger.e('[SocketService] Listener error for $event: $e');
          }
        }
      });
    }
  }

  void off(String event, Function callback) {
    _logger.d('[SocketService] Unregister listener: $event');
    if (_listeners.containsKey(event)) {
      _listeners[event]!.remove(callback);
      if (_listeners[event]!.isEmpty) {
        _listeners.remove(event);
        if (_socket != null) {
          _socket!.off(event);
        }
      }
    }
  }

  void disconnect() {
    _logger.d('[SocketService] Disconnect requested');
    _authSuccess = false;
    _currentToken = null;
    _socket?.disconnect();
    _socket = null;
  }

  Future<void> reconnect() async {
    _logger.d('[SocketService] Reconnect requested');
    disconnect();
    await connect();
  }
}
